### DISCUSSION TITLE
BRouter (offline) use distant waypoints for off-road routes

### DISCUSSION THREAD

**Discussion by abdullahO2** [devemux86/cruiser#204] [Q&A] [answered]
Greetings,

I wanted to share a suggestion for improving the routing feature. Currently, when a user selects a routing endpoint to a location on the map that does not have any roads, the application displays an error message such as "BRouter: to-position not mapped in existing datafile" or "GraphHopper: Waypoint 2 not found." 
And I suggest that the application automatically routing users to the nearest road and then draw a straight line to the endpoint, without requiring user intervention. And I suggest adding a button in the routing settings to enable and disable this feature, And making it available for all routing engines.

While I am aware that BRouter has a "straight line" feature that accomplishes this, it can be challenging for some users, particularly the elderly and non-experts, to access this feature as it requires a lengthy explanation. We believe that incorporating this feature into the application and making it easily accessible through a button in the routing settings would greatly benefit all users, regardless of their level of expertise. and Here is a GIF from Willy explaining the cool feature:
![Place route(way)points on street](https://github.com/devemux86/cruiser/assets/61907628/f4d69176-ed51-441b-a953-fd77c65b24f6)
and Here are clips demonstrating this feature in other applications and drawing a direct line without user interaction:
https://youtu.be/yvXSpP_dkdQ
https://youtu.be/1dD6SDb0Z6Q
https://youtu.be/3Gu9WlyYLRc
This feature is especially crucial for us in our country, where many places are not accessible by roads, but it is still possible to drive a car through them. Incorporating this feature into the application would make a significant difference to users, and we hope you will consider our suggestion.

Thank you for your attention to this matter.
Best regards,
Abdullah Abdulrahman

**Comment by devemux86** [comment]
@abdullahO2 thanks for the interesting idea!

BRouter 1.7 should work much better at attaching the waypoints on the road network.

**Reply by abdullahO2** [reply]
@devemux86 Could a better solution be found by developing a feature (to be optional) such that if a message appears:
(BRouter: to-position not mapped in existing datafile)
The application sets the point to the nearest way and then completes the rest of the route line directly (Segment profile = straight line), or in another way to make the process in the video work automatically:
https://www.youtube.com/watch?v=ketZyL-vUCU

**Reply by devemux86** [reply]
@abdullahO2 thanks for the suggestion.

It is in my list to check if / how it could be done combined with latest BRouter. :slightly_smiling_face:

**Reply by abdullahO2** [reply]
Thank you very much for your interest

**Comment by abdullahO2** [comment]
@devemux86 Thank you for your response. I did try increasing the "Waypoints search range" but 1 kilometer is not enough because sometimes we need to go 20 kilometers inside the desert, or you may need to go directly from point A to point B because there are no roads (some applications have implemented a feature where point A moves according to GPS while point B remains fixed)

**Reply by devemux86** [reply]
@abdullahO2 

> I suggest that the application automatically routing users to the nearest road and then draw a straight line to the endpoint, without requiring user intervention.

You mean adding automatically 2 waypoints:
one at the clicked location and one at the nearest road and connect them with a straight line.

Indeed, this an interesting idea!

**Reply by abdullahO2** [reply]
Right and thanks for your interest

**Comment by 0709wiwiwi** [comment]
@abdullahO2
@devemux86 

- A surprising but very well considered positive suggestion.

Chances of success ?

- You should not set the snap to road range higher but on the contrary lower than is usually set.
- Without nearby snapping road, the routing **automatically** should switch to a beeline connection.
- It than sure must preferably also include a fine usable matching clock guiding TTS notation. 
- Export and import transfers using a simple Cruiser gpx extension is definitely acceptable.
- [Beeline guiding instruction](https://github.com/devemux86/cruiser/discussions/79#discussioncomment-5535928)
  
Imo this newest suggestion is valuable and worth to consider and to test.

**Reply by 0709wiwiwi** [reply]
Yet what follows is better not as suggested than implemented.

_"I suggest that the application automatically routing users to the nearest road and then draw a straight line to the endpoint"._

What if between the nearest road and the endpoint (route waypoint) there is a very large wall or mountain range in the way ? 

If the endpoint (route waypoint) is outside the snap range then the straight line connection should be drawn from the last still snapable shaping point and then further as a straight line directly drawn to the next non snapable route planner point. 

**By the way, a straight line connection stands out very well because of a different color presentation.**
If you don't want that differently colored Straight Line section work a little more precisely when scheduling.

Constantly increasing the snapping range to still get a result is cheating.  It just encourages users to place the planner points very imprecisely and just click away too randomly.  A deviation of 500m may be acceptable for motorcyclists, but for pedestrians, the planning points must be placed more precise.

Different use ? = Different Presets.

Translated with www.DeepL.com/Translator (free version)

**Reply by devemux86** [reply]
@0709wiwiwi indeed automatic waypoints are never safe,
they can be placed anywhere, especially with large ranges.

It is a similar issue with the automatically added waypoints when using gpx files.
In so many cases they are not placed correctly, leading to unnecessary u-turns.

**Reply by abdullahO2** [reply]
@0709wiwiwi 
Thank you for your interest. While it is true that navigating through areas without roads can present challenges, traversing through the desert can be a bit different. In fact, it is often possible to bypass many obstacles and travel in a straight or semi-straight line. To better illustrate my point, I suggest taking a look at some videos that demonstrate this approach:
https://youtu.be/DM3FcttCRIk?t=2945
https://youtu.be/rPfiTWxUlAE?t=3262
https://youtu.be/A7afbEuo2CQ?t=1551
Someone may ask why the roads are not drawn on the map, and the answer is that these roads are too many and some of them are difficult to draw via satellite images. Also, some of these roads can disappear under the sand in a short time, and therefore the next person who comes to the area will create a new road. This happens very often, especially in the sandy areas that make up a very large area of our country.

**Reply by 0709wiwiwi** [reply]
@abdullahO2 

I think I understand what you want to obtain in your big "playground" ;-)
And I especially so have a look at what this could lead to locally then.

- Situation: 
  You place a scheduler point which falls outside the BRouter routing snap range. 

- Current message:   
  "BRouter position not mapped in existing datafile".
   You need to manually place the "Beeline" section.

- Your Proposal: 
   The "Beeline" section is placed automatically.

Practically, I expect this.

- Your environment: Beelines are often needed.
  Concerns quite long manual sections in locations with quite few osm roads.
  Even with 1km snap range, the chances of "unwanted" being wrongly snapped are quite low.
  If an "unwanted" snap does occur you still than need to manually trigger the "beeline".

- My environment: Beelines are rarely needed.
  This concerns than fairly short manual sections at locations with an extensive osm road network.
  The chances of "false snaps" using the 250m (default) set  remains quite high.
  Even with smaller snap ranges, you will often need to manually trigger a planned "beeline" 


- Advantage: 
 The message BRouter position not mapped in existing datafile disappears completely.
 The "automatic" placed Beeline section COLOUR shows no "osm routing" was possible.

_I hope my summary is correct_

**Reply by abdullahO2** [reply]
@0709wiwiwi Yes, your summary is correct

**Comment by devemux86** [comment]
@abdullahO2 

Can you please post an example route and how you propose it to work?
It will help the discussion and everyone understand the point placement.

**Reply by devemux86** [reply]
The animated gif above shows a rather simple example with route extension at end that can easily be solved by the computer.
But routes can be more complicated and waypoints added in the middle, so it's interesting to see what everyone has in mind. :slightly_smiling_face:

**Reply by abdullahO2** [reply]
normal routing - straight line:
https://www.youtube.com/watch?v=j6CvC8TmpHY
straight line - normal routing - straight line:
https://www.youtube.com/watch?v=ketZyL-vUCU

**Reply by devemux86** [reply]
@abdullahO2 thanks very much for the videos!

I now understand your request better and will look into it. :slightly_smiling_face:

**Comment by devemux86** [comment]
@afischerdev has done some work in BRouter. :+1: 

You can see the interesting Cruiser result also here:
- https://github.com/abrensch/brouter/issues/387#issuecomment-2288615917

<img src="https://github.com/user-attachments/assets/065c4d1a-a200-45cf-97d9-ed3d3a876e17" width="50%"/><img src="https://github.com/user-attachments/assets/2ba7c2b0-b9c8-4fa4-831c-ad5b1dbddae5" width="50%"/>

**Reply by Ciutnik** [reply]
Interesting! A couple of times, when trying to calculate a route to a place whose distance from the road network exceeds the limit (specified in the Cruiser settings), I received a corresponding message about the inability to calculate the route. 

Does the above message mean that this will no longer happen, because a straight line will be automatycznie determined for this section?

**Reply by devemux86** [reply]
@Ciutnik 

This is a BRouter function (also in [website](https://brouter.de/brouter-web/)) about the snap of waypoints to the nearest road network.

You can control it in "**Settings | Routing | BRouter (Offline) | Advanced | Waypoints search range**".

> Does the above message mean that this will no longer happen, because a straight line will be automatycznie determined for this section?

Probably, but these straight lines generate extra turn instructions, see above the [2nd image](https://github.com/devemux86/cruiser/discussions/204#discussioncomment-10336976).
So it could be via an option.

**Reply by abdullahO2** [reply]
@devemux86

Thank you so much for the update! I really appreciate @afischerdev's work on this feature in BRouter. I've checked the link you provided, and the results look very promising. This is a significant step towards making off-road navigation easier and more user-friendly.

Thank you again for your attention to this matter and for keeping me in the loop. Your dedication to improving Cruiser is truly commendable.

Best regards,

Abdullah Abdulrahman

**Comment by devemux86** [comment]
This feature is available for testing in the BRouter **test** website, for instructions see here:

- https://github.com/abrensch/brouter/pull/729

---

@abdullahO2 

Can you test it?
This should work on start and end points and also on intermediate points.

**Reply by abdullahO2** [reply]
@devemux86

Thank you for the update! I've tested the new dynamic range feature on the BRouter test website, and it works perfectly for start and end points, as well as intermediate points. It's exactly what I was hoping for, and I'm very impressed with the results.

I'm now eagerly awaiting the updated BRouter app so I can test this functionality in a real-world setting. I'm confident that it will significantly improve the off-road navigation experience in Cruiser.

Thank you again to you and @afischerdev for your continued dedication to making BRouter and Cruiser the best they can be.

Best regards,

Abdullah Abdulrahman

**Reply by devemux86** [reply]
@abdullahO2 thanks for the tests!

I will include this new feature in one of the next Cruiser updates.

**Comment by devemux86** [comment]
BRouter distant waypoints for off-road routes are available in the latest Cruiser.

It allows setting waypoints away from the road network for off-road routes.
It connects the distant waypoints to the road network with straight lines leading to successful route calculation.

It currently requires calculation of the whole route to work properly.

BRouter discussion with the reports:
- https://github.com/abrensch/brouter/issues/387

The option is available in "Settings | Routing | BRouter (offline) | Advanced | Use distant waypoints":

<img src="https://github.com/user-attachments/assets/8a3a5e07-a2b9-4fcf-a19d-f7b4d2beed61" width="50%"/><img src="https://github.com/user-attachments/assets/f893c671-f888-4afb-9868-318e2fde7021" width="50%"/>

**Comment by abdullahO2** [comment]
@devemux86

Thank you so much for your great efforts in adding the "Use distant waypoints" feature to the Cruiser app. I have been following the development of this feature since its inception, and I am very happy to see that it is finally available. This feature will make a big difference in the navigation experience for users who rely on Cruiser, especially in desert areas or off-road.

I appreciate your attention to user needs and your continued cooperation in improving the Cruiser app. I also thank @afischerdev for the hard work in developing this feature in BRouter. Your dedication to making Cruiser the best navigation application is truly commendable.

Best regards,

Abdullah Abdulrahman

**Comment by devemux86** [comment]
BRouter distant waypoints for off-road routes is available in **[Cruiser 5.1.0-b6](https://github.com/devemux86/cruiser/discussions/671)**.
(see the instructions [here](https://github.com/devemux86/cruiser/discussions/204#discussioncomment-11284031))

---

The turn instructions are generated by the router.

@abdullahO2 and everyone else, please test this new BRouter feature.

You can report here or in the BRouter repository with [BRouter website](https://brouter.de/brouter-web/) route links:
- https://github.com/abrensch/brouter/issues/387

**Reply by Ciutnik** [reply]
At first glance it works great! Thanks a lot 👍

Maybe the only thing missing is the ability to connect several off-road points with a straight line. I mean, for example, this case (Via1 → End):

<img src="https://github.com/user-attachments/assets/2bb46b1c-ef41-4662-804a-757418e1d357" width="50%"/>

**Reply by devemux86** [reply]
@Ciutnik For such cases it is better to use the "Straight line" profile in the waypoints.
Also you can suggest it to the [BRouter issue](https://github.com/abrensch/brouter/issues/387) with a BRouter website link.

Note: BRouter calculates routes with intermediate points separately segment by segment and then joins them.

Studying the BRouter code (@afischerdev can explain better):
BRouter increases the search range of waypoints and tries to connect them with the nearest road network.

**Reply by Ciutnik** [reply]
@devemux86 
> For such cases it is better to use the "Straight line" profile in the waypoints.

Right! I completely forgot about that option! In such special cases, that does the trick 👍

<img src="https://github.com/user-attachments/assets/3b47214a-6929-4aba-a5ca-ef5ae77cbbd8" width="50%"/>

**Reply by Ciutnik** [reply]
Looking at my screenshot I think the "connector" color might be the same as the "Straight Line" profile. But I know that it takes the color defined in the profile used, so it's not that obvious (and maybe not easy to implement) 🙂

**Reply by devemux86** [reply]
@Ciutnik Indeed, as this is not a straight line profile between waypoints with its own color.
It is still part of the normal route and has the color of its profile, which could be anything.

**Comment by devemux86** [comment]
You can also test this feature in the [BRouter **test** website](https://brouter.de/brouter-test/).

(as the main BRouter website hasn't included it yet)

You just need to add this constant in the profile:
`assign use_dynamic_range   = true  # %use_dynamic_range% | Enable distant start/end points | boolean`

(Cruiser driving profile already includes it)

**Comment by devemux86** [comment, marked as answer]
- BRouter (offline) use distant waypoints for off-road routes is available in https://github.com/devemux86/cruiser/discussions/684

**Comment by abdullahO2** [comment]
@devemux86

Thank you so much for implementing this feature in Cruiser 5.1.1! I really appreciate your dedication and hard work in bringing this to fruition.

The latest changes in BRouter (specifically those in [PR #747](https://github.com/abrensch/brouter/pull/747)) have made this feature work exceptionally well. It's a game-changer for off-road navigation, and I'm incredibly impressed with the results. Thank you again for making this possible!

**Comment by devemux86** [comment]
> The latest changes in BRouter (specifically those in https://github.com/abrensch/brouter/pull/747) have made this feature work exceptionally well.

They are already included in **[Cruiser 5.2.0](https://github.com/devemux86/cruiser/discussions/701)**.

@abdullahO2 Thanks for your support!

**Comment by Ciutnik** [comment]
Does enabling this feature still exclude intelligent route calculation, i.e. only the new/changed section? If so, maybe it's worth adding some warning on the second line, below the switch?

**Reply by devemux86** [reply]
It calculates (internally) the whole route so the BRouter algorithms can work their magic. :slightly_smiling_face: 
We can add an extra description in the setting.

**Reply by Ciutnik** [reply]
> We can add an extra description in the setting.

Unfortunately, I had to disable this function due to the drastic increase in the time required to recalculate a route consisting of several VIA points. There's a price to pay, but it would be good if users knew what the cost of enabling this option is 🙂
